ceph: add multus support (#4077)
rook/rook: Documentation/ceph-cluster-crd.md

      Recommended:
    * If you have a single Rook Ceph cluster, set the `rulesNamespace` to the same namespace as the cluster or keep it empty.
    * If you have multiple Rook Ceph clusters in the same Kubernetes cluster, choose the same namespace to set `rulesNamespace` for all the clusters (ideally, namespace with prometheus deployed). Otherwise, you will get duplicate alerts with duplicate alert definitions.

> * `network`: The network settings for the cluster, refer to [network configuration settings](#network-configuration-settings)

Reviewer: ```suggestion
* `network`: For the network settings for the cluster refer to the [network configuration settings](#network-configuration-settings)
```

* `pg_autoscaler`: Rook will configure all new pools with PG autoscaling by setting: `osd_pool_default_pg_autoscale_mode = on`

### Network Configuration Settings

Configure the network that will be enabled for the cluster and services. This is optional and if not defined then the default SDN will be used.

> * `provider`: Specifies the network provider that will be used to connect the network interface. You can choose between `host`, and `multus`.

Reviewer: What if not specified? You get the default sdn?

Author: Yes.

Reviewer: How about if the comment above about the default be moved to this line?

Reviewer: Need a comment about multus being expermental

* `pg_autoscaler`: Rook will configure all new pools with PG autoscaling by setting: `osd_pool_default_pg_autoscale_mode = on`

### Network Configuration Settings

> Configure the network that will be enabled for the cluster and services. This is optional and if not defined then the default SDN will be used.

Reviewer: ```suggestion
Configure the network that will be enabled for the cluster and services. If not specified, the default SDN will be used.
```

* `provider`: Specifies the network provider that will be used to connect the network interface. You can choose between `host`, and `multus`.
* `selectors`: List the network selector(s) that will be used associated by a key.

> The selector keys are hard-coded to `public` and `cluster`.

Reviewer: ```suggestion
The selector keys are required to be `public` and `cluster`.
```

Reviewer: I'd prefer "required" instead of "hard-coded" everywhere it is mentioned. "hard-coded" is more of an implementation detail IMO.

Based on the configuration, the operator will do the following:

>   1. if only the `public` selector is specified both `public_network` and `cluster_network` Ceph settings will listen on that interface

Reviewer: Instead of mentioning the name of the ceph setting, what if we tell them about the Ceph behavior? We could include a link to the Ceph networking docs if they want to know more about the Ceph settings.

  2. if both `public` and `cluster` selectors are specified the first one will point to `public_network` flag and the second one to `cluster_network`

In order to work, each selector value must match a `NetworkAttachmentDefinition` object in Multus.

> The following are just arbitrary examples and not hardcoded names.

Reviewer: ```suggestion
The following are just arbitrary examples and not required names.
```

Reviewer: wait, aren't `public` and `cluster` still required? Or you mean the names that come in the mapping aren't required.

Author: The value, clarified.
